### PR TITLE
Add support for Duration type

### PR DESCRIPTION
**Pull request by Ponewor**
Fixes: #320 

This works great with Scylla but for some reason it fails with Cassandra (at least with 4.0.1). The column type code that Cassandra is sending for Duration is 0 instead of 0x15 and it is followed with type's name: "org.apache.cassandra.db.marshal.DurationType". How should we deal with it?

Apart from that (and a few missing tests or docstrings), is there anything serious missing here? 

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Comment by psarna** [comment]
@Ponewor I have no idea why Cassandra decided to encode duration as a custom type instead of a native type, but it is legal in CQL v4. Here's an excerpt from https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec:
```
The option for <type> is either a native
          type (see below), in which case the option has no value, or a
          'custom' type, in which case the value is a [string] representing
          the fully qualified class name of the type represented. Valid option
          ids are:
            0x0000    Custom: the value is a [string], see above.
            0x0001    Ascii
            0x0002    Bigint
            0x0003    Blob
            (...)
```

So this is actually a mild compatibility bug - not a serious one, because our driver should be free to not recognize specific custom types, but perhaps we should have a map of handlers for a few known ones - e.g. `org.apache.cassandra.db.marshal.DurationType`.

I created https://github.com/scylladb/scylla-rust-driver/issues/364 to track the problem. For now, either make the test optional or add a Cassandra-specific bypass - namely, let the test pass either when it returns a correct duration, or when it fails because it received a custom type. If you decide to go with the latter, please also add a comment about it in the test case.

File: scylla/src/frame/value.rs [hunk 1]
```diff
@@ -410,6 +418,82 @@ impl Value for Counter {
...
+    assert_eq!(zig_zag_encode(-3), 5);
+    assert_eq!(zig_zag_encode(3), 6);
+}
+
+#[test]
+fn vint_encode_test() {
+    let mut buf: Vec<u8> = Vec::with_capacity(128);
+    let n: i64 = 256000;
+    vint_encode(n, &mut buf);
+    assert_eq!(buf, vec![199, 208, 0])
```
Comment (piodul) [inline scylla/src/frame/value.rs:477]: I think we need more comprehensive tests for (unsigned) vint serialization and deserialization:

- Do the above check with some more numbers with different lengths,
  - Include corner cases - 0 and maxint,
- For each bit length of a number, choose a number of such length and check if serialization then deserialization returns the original number.

File: scylla/src/frame/value_tests.rs [hunk 2]
```diff
@@ -718,3 +719,53 @@ async fn test_cqlvalue_udt() {
...
+        .unwrap()
+        .rows
+        .unwrap();
+
+    assert_eq!(rows.len(), 1);
+    assert_eq!(rows[0].columns.len(), 1);
+
+    let received_duration_cql_value = rows[0].columns[0].as_ref().unwrap();
+
+    assert_eq!(received_duration_cql_value, &duration_cql_value);
```
Comment (piodul) [inline scylla/src/frame/value_tests.rs:770]: In this test, please also check that a duration inserted with a CQL query in which the duration is encoded in the CQL string is then deserialized properly by the driver.

File: scylla/src/frame/response/result.rs [hunk 3]
```diff
@@ -651,6 +662,23 @@ fn deser_cql_value(typ: &ColumnType, buf: &mut &[u8]) -> StdResult<CqlValue, Par
...
+            let (months, offset) = vint_decode(buf);
+            buf.advance(offset);
+            let (days, offset) = vint_decode(buf);
+            buf.advance(offset);
+            let (nanoseconds, offset) = vint_decode(buf);
+            buf.advance(offset);
+
+            CqlValue::Duration(CqlDuration {
+                months: months as i32,
+                days: days as i32,
```
Comment (piodul) [inline scylla/src/frame/response/result.rs:678]: It would be good to check that the decoded `i64`'s do not overflow.

File: scylla/src/frame/value.rs [hunk 4]
```diff
@@ -410,6 +418,82 @@ impl Value for Counter {
...
+pub(crate) fn vint_decode(buf: &[u8]) -> (i64, usize) {
+    let extra_bytes = buf[0].leading_ones() as usize;
+    let mut v = 0;
+
+    for (i, byte) in buf.iter().enumerate().take(extra_bytes + 1).skip(1) {
+        v += (*byte as u64) << (8 * (extra_bytes - i) as u64);
+    }
+
+    if extra_bytes != 8 {
+        v += ((buf[0] - !((0xff >> extra_bytes) as u8)) as u64) << (8 * extra_bytes);
```
Comment (piodul) [inline scylla/src/frame/value.rs:455]: The combination of parentheses, prefix (`!`) and postfix (`as`) operators make it hard to parse for me. Will the following work?

```rust
let first_byte_bits = buf[0] & (0xffu8 >> extra_bytes);
v |= first_byte_bits as u64 << (8 * extra_bytes);
```

File: scylla/src/frame/value.rs [hunk 5]
```diff
@@ -410,6 +418,82 @@ impl Value for Counter {
...
+fn vint_encode(v: i64, buf: &mut Vec<u8>) {
+    let mut v = zig_zag_encode(v);
+    let number_of_bytes = 8 - v.leading_zeros() / 8;
+
+    if number_of_bytes != 8 {
+        let mut extra_bytes = number_of_bytes - 1;
+        if number_of_bytes > v.leading_zeros() % 8 {
+            extra_bytes += 1;
+        }
+        v += (!((0xff >> extra_bytes) as u8) as u64) << (8 * extra_bytes);
```
Comment (piodul) [inline scylla/src/frame/value.rs:438]: How about splitting it into two parts:

```rust
let length_bits = !(0xffu8 >> extra_bytes);
v |= length_bits as u64 << (8 * extra_bytes);
```

**Comment by piodul** [comment]
@Ponewor The Duration type was introduced in v5 of the CQL protocol. There was a need for this type to be supported in Scylla - however, Scylla only supports v4 so the support for it was backported to the older version of the protocol.

If you look at the protocol specifications as defined in the Cassandra repository, you can see that in v4 there is no concept of Duration ([see here](https://github.com/apache/cassandra/blob/185893256f10c14207bffe49ae733fb1a970aec5/doc/native_protocol_v4.spec#L636-L637)), it is only introduced in v5 ([here](https://github.com/apache/cassandra/blob/185893256f10c14207bffe49ae733fb1a970aec5/doc/native_protocol_v5.spec#L841-L843)).

In order to be compatible with the protocol v4, I guess that Cassandra decides to encode Durations as a "Custom" type. I'm not aware of how the encoding looks, though. I agree with @psarna on how to handle this for Cassandra.

**Review by piodul — CHANGES_REQUESTED** [review]

**Comment by psarna** [comment]
> @Ponewor The Duration type was introduced in v5 of the CQL protocol. There was a need for this type to be supported in Scylla - however, Scylla only supports v4 so the support for it was backported to the older version of the protocol.

Ah, makes sense. Btw, we could consider allowing the user to declare the version explicitly to allow "version spoofing". That would allow users to set v5 here, e.g. if they know that they'll only use the subset of v5 that's actually implemented. It would act as a trivial workaround for issues like this.

> 
> If you look at the protocol specifications as defined in the Cassandra repository, you can see that in v4 there is no concept of Duration ([see here](https://github.com/apache/cassandra/blob/185893256f10c14207bffe49ae733fb1a970aec5/doc/native_protocol_v4.spec#L636-L637)), it is only introduced in v5 ([here](https://github.com/apache/cassandra/blob/185893256f10c14207bffe49ae733fb1a970aec5/doc/native_protocol_v5.spec#L841-L843)).
> 
> In order to be compatible with the protocol v4, I guess that Cassandra decides to encode Durations as a "Custom" type. I'm not aware of how the encoding looks, though. I agree with @psarna on how to handle this for Cassandra.

**Comment by psarna** [comment]
any updates on this one?

File: scylla/src/frame/response/result.rs [hunk 6]
```diff
@@ -1,12 +1,13 @@
...
+use crate::frame::types::vint_decode;
+use crate::frame::value::{Counter, CqlDuration};
 use crate::frame::{frame_errors::ParseError, types};
 use bigdecimal::BigDecimal;
 use byteorder::{BigEndian, ReadBytesExt};
 use bytes::{Buf, Bytes};
+use chrono;
 use chrono::prelude::*;
-use chrono::Duration;
 use num_bigint::BigInt;
```
Comment (piodul) [inline scylla/src/frame/response/result.rs:11]: Changing the `result.rs` file to use the fully-qualified `chrono::Duration` instead of importing the `Duration` type directly is a refactor which should be placed in its separate commit.

File: scylla/src/frame/types.rs [hunk 7]
```diff
@@ -525,3 +525,690 @@ fn type_inet() {
...
+    assert_eq!(zig_zag_decode(2), 1);
+    assert_eq!(zig_zag_decode(3), -2);
+    assert_eq!(zig_zag_decode(4), 2);
+    assert_eq!(zig_zag_decode(5), -3);
+    assert_eq!(zig_zag_decode(6), 3);
+}
+
+#[test]
+fn unsigned_vint_encode_and_decode_test() {
+    let unsigned_vint_encoding = vec![
```
Comment (piodul) [inline scylla/src/frame/types.rs:605]: Could we auto-generate at least some of those test cases?

File: scylla/src/frame/value_tests.rs [hunk 8]
```diff
@@ -723,3 +724,156 @@ async fn test_cqlvalue_udt() {
...
+
+    session
+        .query(
+            "INSERT INTO cqlvalue_duration_test (k, my) VALUES (5, ?)",
+            (&duration_cql_value,),
+        )
+        .await
+        .unwrap();
+
+    session
```
Comment (piodul) [inline scylla/src/frame/value_tests.rs:768]: I think we can make this test shorter if you declare a list of strings, loop over them, construct appropriate query strings and `.query` them.

File: scylla/src/frame/value_tests.rs [hunk 9]
```diff
@@ -723,3 +724,156 @@ async fn test_cqlvalue_udt() {
...
+        .unwrap();
+
+    assert_eq!(rows.len(), 1);
+    assert_eq!(rows[0].columns.len(), 1);
+
+    let received_duration_cql_value = rows[0].columns[0].as_ref().unwrap();
+
+    assert_eq!(received_duration_cql_value, &duration_cql_value);
+
+    let rows = session
```
Comment (piodul) [inline scylla/src/frame/value_tests.rs:809]: Instead of fetching every row in a separate select, you can run a single select and fetch all rows from the table. If you want to have them in a sorted order, then you can use a single partition and use different clustering keys.

**Review by piodul — CHANGES_REQUESTED** [review]

File: scylla/src/frame/types.rs (hunk 7 above)
Comment (Ponewor) [reply scylla/src/frame/types.rs:605]: But the encodings need to be hardcoded anyway, don't they? There's almost nothing to gain in terms of LoCs.

How these cases were generated:
1. Python script generated test cases in Go
2. Go program with vint encoding logic copied from Go Driver took generated test cases and printed the encodings.
3. Python script took the output and generated Rust test cases.

Comment (piodul) [reply scylla/src/frame/types.rs:605]: OK, it makes sense to keep them spelled out. Let's keep them. The tests cases in `vint_encode_and_decode_test` should definitely be autogenerated, though.

**Review by piodul — APPROVED** [review]
